org.civicrm.reqangular/test/services/api.spec.js

```diff
@@ -99,26 +103,84 @@ define([
         });
       });
 
+      describe('caching', function () {
+        var sampleResult = 'sample result';
+        var sampleParams = {};
+
+        beforeEach(function () {
+          spyOn($cacheFactory.get('$http'), 'remove').and.callThrough();
+        });
+
+        describe('when "returnCachedData" parameter is not passed', function () {
+          beforeEach(function () {
+            expectAndSendGET(sampleResult, sampleParams);
+            $rootScope.$digest();
+          });
+
+          it('does not flush cache before HTTP call', function () {
+            expect($cacheFactory.get('$http').remove).not.toHaveBeenCalled();
+          });
+
+          it('caches data by default', function () {
+            expect($cacheFactory.get('$http').info().size).toBe(1);
+          });
+        });
+
+        describe('when "returnCachedData" parameter passed as TRUE', function () {
+          beforeEach(function () {
+            expectAndSendGET(sampleResult, sampleParams, true);
+            $rootScope.$digest();
+          });
+
+          it('does not flush cache before HTTP call', function () {
+            expect($cacheFactory.get('$http').remove).not.toHaveBeenCalled();
+          });
+
+          it('caches data', function () {
+            expect($cacheFactory.get('$http').info().size).toBe(1);
+          });
+        });
+
+        describe('when "returnCachedData" parameter passed as FALSE', function () {
+          beforeEach(function () {
+            expectAndSendGET(sampleResult, sampleParams, false);
+            $rootScope.$digest();
+          });
+
+          it('flushes cache before HTTP call', function () {
+            expect($cacheFactory.get('$http').remove).toHaveBeenCalledWith(
+              '/civicrm/ajax/rest?action=action&entity=entity&json=%7B%22options%22:%7B%22limit%22:0%7D%7D&sequential=1');
+          });
+
+          it('caches updated data for future requests', function () {
+            expect($cacheFactory.get('$http').info().size).toBe(1);
+          });
+        });
+      });
+
       /**
        * Mocks and sends a fake GET request
        *
        * @param  {any} returnValue - value to be returned by the GET request
        * @param  {Object} params - params to be used in the API call
+       * @param  {Boolean} cache
        * @return {Promise}
        */
-      function expectAndSendGET (returnValue, params) {
+      function expectAndSendGET (returnValue, params, cache) {
         $httpBackend
           .whenGET(new RegExp('action=' + action + '&entity=' + entity))
           .respond(returnValue);
 
-        return api.sendGET(entity, action, params, true);
+        return api.sendGET(entity, action, params, cache);
       }
     });
 
-    describe('sendPOST', function () {
+    describe('sendPOST()', function () {
       var promise;
 
-      afterEach(function () { $httpBackend.flush(); });
+      afterEach(function () {
+        $httpBackend.flush();
```

Comment:
we just discussed that we shouldn't do this and use a different format?

Reply:
Done. Moreover, these ones were not working at all (passing when they should not pass):

```js
$httpBackend.expectPOST('/civicrm/ajax/rest', $httpParamSerializer({
              json: { options: { limit: limit } },
              sequential: 1,
              action: action,
              entity: entity
            }));
```

Now they are fixed.